src/qibolab/native.py

```diff
@@ -188,6 +205,10 @@ def from_dict(cls, qubit, native_gates):
         pulses = {n: NativePulse.from_dict(n, pulse, qubit=qubit) for n, pulse in native_gates.items()}
         return cls(**pulses)
 
+    @property
+    def to_dict(self):
+        return {fld.name: getattr(self, fld.name).to_dict for fld in fields(self)}
```

Comment:
> However, in this case why `asdict()` is not enough?

An I had with `asdict()` was that some of these objects contain cyclic dependencies and this leads to an infinite recursion. For example `Qubit` has native gates, but each native gate also has a reference back to the `Qubit`.
Another issue with `asdict()` is that it is a pain to exclude a field from it, which still happens very often in our case (if you see most `to_dict`s implemented in this PR have some kinds of exclusions).

I am fine with Pydantic, even though I don't have any experience using it, but I am not sure if it will solve these two problems. Unless we manually try to clean up the objects and their relations to avoid cycles and only have the fields we need.

By the way, I think `asdict()` is acting recursively by default if a dataclass field is another dataclass. At least that was the behavior in some toy examples I tried. I think this is causing the infinite recursion.